### DISCUSSION TITLE
I need to write a small script

### DISCUSSION THREAD

**Discussion by nupharr** [nushell/nushell#13050] [Q&A] [answered]
I need a script that will automatically activate the python virtual environment (yes, I know that there is a ready-made auto-venv, but for some reason it does not work correctly for me). I'm new to nushell and it will be difficult for me to do this on my own at this stage. I would be grateful if someone helps.

Script description:
When moving to a new directory, the script automatically checks for the presence of the ".venv" folder, if there is one, it automatically activates the virtual environment and returns to the initial directory, as shown in the code
```nu
cd ./.venv/Script/
overlay use activate.nu
cd ../..
```

If there is a ".venv" folder but it is not possible to activate the virtual environment, then the message "Failed to activate the Python virtual environment!" is displayed on the screen.

I use windows 10

**Comment by ghost** [comment]
its not a script if it should execute automatically whenever something happens, but instead a part of your `~/.config/nushell/config.nu`.

for what you want to do you'll need [hooks](https://www.nushell.sh/book/hooks.html).

and you'll end up with something like this:

```nu
$env.config = ($env.config | upsert hooks.env_change.PWD {|config|
    let val = ($config | get -i hooks.env_change.PWD)
    let code = {|before, after| if ($"($after)/.venv" | path exists) {overlay use .venv/Script/activate.nu}}

    if $val == null {
        $val | append $code
    } else {
        [$code]
    }
})
```

(this does not trigger via `cd ./.venv/Script/`, but instead whenever your new directory contains a `.venv` since it makes no sense to use automation otherwise and you could just manually run it)

>  the message "Failed to activate the Python virtual environment!" is displayed on the screen.

i for my part have given up on the "official" nushell python venv implementations and run my own one (<https://github.com/Jan9103/nushell_python/blob/main/pyvenv.nu>), but maybe someone else has solved the problem in a proper way

**Reply by nupharr** [reply]
I get an error when I add this code to my config file
![image](https://github.com/nushell/nushell/assets/72264744/acbda098-2dfc-4113-a5aa-d6b00013151b)

I asked to use `cd ./.venv/Scripts` because I can't activate the virtual environment via `overlay use` if the file is not in the directory I'm in
![image](https://github.com/nushell/nushell/assets/72264744/50a99e7e-921f-45b9-98e2-273848f28560)
how can I fix this error?

**Reply by ghost** [reply, marked as answer]
> I asked to use cd ./.venv/Scripts because I can't activate the virtual environment via overlay use if the file is not in the directory I'm in

in that case changing the code line to
```nu
let code = {|before, after| if ($"($after)/.venv" | path exists) {cd .venv/Script; overlay use activate.nu: cd ../.. }}
```
should fix it (switch to the directory and back)

> how can I fix this error?

if the error occurs uppon entering a directory containing a `.venv` it appears as if it does not contain a `activate.nu`. you could replace `if ($"($after)/.venv" | path exists)` with `if ($"($after)/.venv/Script/activate.nu" | path exists)` to only run the code when it actually exists, but i suspect it will never do anything.  
maybe theres a issue with the the file not getting created or something. in that case my script might be a solution since it just emulates the behavior without relying on anything external.

if the error occurs uppon opening nushell its probably the old issue of "a file has to exist before the code is loaded if you want to use its code"..  
maybe it can be fixed by putting the `overlay load` into a custom function and then calling the function or some other weird workaround.  
however its probably easier (and also safer tbh) to use `def pyvenv [] {cd .venv/Script; overlay use activate.nu: cd ../.. }` instead of a hook and just running it manually.

**Comment by MatrixManAtYrService** [comment]
I totally respect the desire to do this in nushell, but just as an FYI, you can achieve something similar with [direnv](https://direnv.net/).  Just create a `.envrc` in the directory with contents:
```
export VIRTUAL_ENV=./venv
export PATH=./venv/bin:$PATH
```

Where `./venv` is the path of the virtual environment relative to the `.envrc` file.
